librz/core/linux_heap_glibc.c

```diff
@@ -126,25 +136,258 @@ static GHT GH(get_main_arena_with_symbol)(RzCore *core, RzDebugMap *map) {
 	return main_arena;
 }
 
+static const ut8* GH(get_glibc_banner)(RzCore *core, const char *section_name,
+									   const char *libc_path) {
+	RzBin *bin = NULL;
+	RzBinFile *current_bf = NULL;
+	if (core->bin) {
+		bin = core->bin;
+		current_bf = rz_bin_cur(bin);
+	} else {
+		bin = rz_bin_new();
+	}
+
+	void **iter;
+	RzBinSection *rz_section = NULL;
+	ut8 *buf = NULL;
+
+	RzBinOptions opt;
+	rz_bin_options_init(&opt, -1, 0, 0, false);
+	opt.obj_opts.elf_load_sections = rz_config_get_b(core->config, "elf.load.sections");
+	opt.obj_opts.elf_checks_sections = rz_config_get_b(core->config, "elf.checks.sections");
+	opt.obj_opts.elf_checks_segments = rz_config_get_b(core->config, "elf.checks.segments");
+
+	RzBinFile *libc_buf = rz_bin_open(bin, libc_path, &opt);
+	if (!libc_buf) {
+		return NULL;
+	}
+
+	RzPVector *sections = rz_bin_object_get_sections(libc_buf->o);
+
+	rz_pvector_foreach (sections, iter) {
+		rz_section = *iter;
+		if (strncmp(rz_section->name, section_name, strlen(section_name))) {
+			continue;
+		}
+		buf = calloc(rz_section->size, 1);
+		GHT read_size = rz_buf_read_at(libc_buf->buf, rz_section->paddr, buf, rz_section->size);
+		if (read_size != rz_section->size) {
+			free(buf);
+			buf = NULL;
+			goto cleanup;
+		}
+		buf = (ut8*)rz_mem_mem((const ut8*)buf, rz_section->size, (const ut8 *)"GNU C Library", strlen ("GNU C Library"));
+	}
+
+cleanup:
+
+	rz_bin_file_delete(bin, libc_buf);
+	rz_bin_file_set_cur_binfile(bin, current_bf);
+	return buf;
+}
+
+RZ_API double GH(rz_get_glibc_version)(RzCore *core, const char *libc_path,
+									   const ut8 *banner_start) {
+	double version = 0.0;
+	
+	if (!banner_start) {
+		banner_start = GH(get_glibc_banner)(core, ".rodata", libc_path);
+		if (!banner_start) {
+			goto error;
+		}
+	}
+
+	const char *pattern = "release version (\\d.\\d\\d)";
+	RzRegex *re = rz_regex_new(pattern, RZ_REGEX_EXTENDED | RZ_REGEX_CASELESS, 0);
+	RzPVector *matches = rz_regex_match_first(re, (const char *)banner_start,
+											  RZ_REGEX_ZERO_TERMINATED, 0 , RZ_REGEX_DEFAULT);
+	RzRegexMatch *match = rz_pvector_at(matches, 1);
+	char *version_str = rz_str_ndup((const char*)banner_start + match->start, match->len);
+	if (!version_str) {
+		rz_pvector_free(matches);
+		return version;
+	}
+	version = strtod(version_str, NULL);
+	rz_pvector_free(matches);
+	if (version != 0) {
+		RZ_LOG_INFO ("libc version %.2f identified from .rodata banner", version);
+		return version;
+	}
+
+error:
+	RZ_LOG_WARN ("get_glibc_version failed");
+	return version;
+}
+
+static RzList *GH(fetch_tls_address_list)(RzDebug *dbg) {
+	RzList *t_list = NULL;
+
+	if (!dbg->threads) {
+		return t_list;
+	}
+
+	RzDebugPid *th;
+	RzListIter *it;
+
+	rz_list_foreach (dbg->threads, it, th) {
```

Comment:
this loop is useful only if `th->tls` is zero. probably we can fail this earlier at line 225 by checking if `tls` is not zero.

Reply:
this also allows to remove the check of `tls` in the loop.

Reply:
I have combined two conditions, but we have to traverse through each thread and check if the `tls` is fetched for that specific thread. lmk if i am missing anything.

Reply:
no, please have this check much earlier so we skip some useless branch instructions within the loop. combine this with line 227 if

Reply:
Refactored this :+1: